### PR TITLE
adding support for multilinugal cache aware model

### PR DESCRIPTION
**Pull request by naymaraq**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [ASR]

# Changelog

- Adds support for prompt-conditioned multilingual models (e.g. nvidia/nemotron-3.5-asr-streaming-0.6b) to the cache-aware RNNT streaming inference pipeline.

# Usage

```python
python asr_streaming_infer.py \
    --config-path=../conf/asr_streaming_inference/ --config-name=cache_aware_rnnt.yaml \
    asr.model_name=nvidia/nemotron-3.5-asr-streaming-0.6b \
    streaming.att_context_size=[56,13] \
    lang=de-DE \ #auto
    audio_file=<manifest.jsonl> output_filename=out.jsonl
```

# Results

### Evaluation setup

| | |
|---|---|
| **Model** | `nvidia/nemotron-3.5-asr-streaming-0.6b` (40 supported locales) |
| **Dataset** | FLEURS test — 35 locales with a matching subset, 26,739 utterances, **86.5 h** |
| **Metric** | **CER**, not WER — the set includes Japanese, Chinese and Thai, which have no word boundaries |
| **Normalization** | punctuation stripped → lowercased → whitespace collapsed → compared character-by-character (`jiwer.process_characters`; identical to `wer_jiwer.py --cer`, mode `no_pc`) |
| **Scoring** | corpus-level, **as produced** — no post-hoc tag stripping on either side |
| **Settings** | `batch_size=128`, `num_slots=128`, `att_context_size=[56,13]`, `warmup_steps=0`, single GPU |
| **Reference** | `examples/asr/asr_cache_aware_streaming/speech_to_text_cache_aware_streaming_infer.py`, run with `strip_lang_tags=true` |

### Column definitions

| column | meaning |
|---|---|
| **ours-fp32** | this PR, `asr.compute_dtype=float32`, explicit locale prompt |
| **ours-bf16** | this PR, `asr.compute_dtype=bfloat16` (pipeline default), explicit locale prompt |
| **ref-pd** | reference script, `pad_and_drop_preencoded=True` — **the apples-to-apples column**: same first-chunk policy as our pipeline |
| **ref-nopd** | reference script, `pad_and_drop_preencoded=False` (its default) — drops no pre-encoded frames on step 0 |
| **\*-auto** | the same four configurations with `lang=auto` (model-detected language) instead of an explicit locale |

The reference is always fp32 — it raises `NotImplementedError` for any other dtype
(`speech_to_text_cache_aware_streaming_infer.py:336-342`), so there is no bf16 reference column.

`*` marks the 5 supported locales that have **no FLEURS subset of their own**, evaluated on their
sibling language's audio (`en-GB` on `en_us` data, and so on). These are paired comparisons where only
the prompt differs — they share audio and reference text with their base row, so they are not
independent samples. Their `auto` columns repeat the base row, since `auto` ignores the locale prompt.

### Table

| language | locale | utts | ours-fp32 | ours-bf16 | ref-pd | ref-nopd | ours-fp32-auto | ours-bf16-auto | ref-pd-auto | ref-nopd-auto |
|---|---|---|---|---|---|---|---|---|---|---|
| it_it | it-IT | 865 | 3.36 | 3.37 | 3.36 | 3.35 | 3.37 | 3.36 | 3.36 | 3.32 |
| es_419 \* | es-ES | 908 | 3.57 | 3.55 | 3.58 | 3.55 | 3.58 | 3.58 | 3.59 | 3.59 |
| es_419 | es-US | 908 | 3.57 | 3.57 | 3.58 | 3.56 | 3.58 | 3.58 | 3.59 | 3.59 |
| tr_tr | tr-TR | 743 | 4.18 | 4.20 | 4.18 | 4.17 | 4.16 | 4.21 | 4.16 | 4.22 |
| uk_ua | uk-UA | 750 | 4.59 | 4.70 | 4.49 | 4.62 | 6.32 | 6.32 | 6.14 | 5.63 |
| pt_br \* | pt-PT | 919 | 4.61 | 4.55 | 4.60 | 4.62 | 4.70 | 4.63 | 4.69 | 4.62 |
| pt_br | pt-BR | 919 | 4.80 | 4.79 | 4.80 | 4.81 | 4.70 | 4.63 | 4.69 | 4.62 |
| de_de | de-DE | 862 | 4.81 | 4.78 | 4.82 | 5.08 | 5.07 | 4.98 | 4.92 | 4.90 |
| ru_ru | ru-RU | 775 | 5.27 | 5.26 | 5.29 | 5.25 | 5.44 | 5.49 | 5.46 | 5.41 |
| fr_fr | fr-FR | 676 | 5.28 | 5.27 | 5.28 | 5.18 | 5.32 | 5.28 | 5.33 | 5.18 |
| fr_fr \* | fr-CA | 676 | 5.42 | 5.36 | 5.34 | 5.25 | 5.32 | 5.28 | 5.33 | 5.18 |
| nl_nl | nl-NL | 364 | 5.50 | 5.71 | 5.65 | 5.51 | 5.77 | 5.71 | 5.80 | 5.58 |
| ar_eg | ar-AR | 428 | 5.95 | 5.94 | 5.94 | 6.03 | 6.03 | 6.01 | 6.02 | 6.13 |
| en_us | en-US | 647 | 5.96 | 5.89 | 5.91 | 5.87 | 6.39 | 6.54 | 6.03 | 6.81 |
| fi_fi | fi-FI | 918 | 6.21 | 6.21 | 6.21 | 6.32 | 6.29 | 6.17 | 6.28 | 6.31 |
| en_us \* | en-GB | 647 | 6.23 | 6.37 | 5.89 | 6.17 | 6.39 | 6.54 | 6.03 | 6.81 |
| bg_bg | bg-BG | 658 | 6.26 | 6.27 | 6.24 | 6.30 | 8.00 | 8.04 | 8.11 | 7.70 |
| hi_in | hi-IN | 418 | 6.51 | 6.46 | 6.58 | 6.58 | 7.23 | 7.29 | 7.28 | 7.78 |
| et_ee | et-EE | 893 | 6.58 | 6.47 | 6.55 | 6.47 | 6.83 | 6.71 | 6.84 | 6.82 |
| nb_no | nb-NO | 357 | 6.64 | 6.61 | 6.65 | 6.71 | 6.61 | 6.60 | 6.63 | 6.67 |
| nb_no \* | nn-NO | 357 | 6.71 | 6.61 | 6.75 | 6.75 | 6.61 | 6.60 | 6.63 | 6.67 |
| pl_pl | pl-PL | 758 | 7.09 | 7.08 | 7.07 | 7.15 | 7.50 | 7.57 | 7.47 | 7.75 |
| sk_sk | sk-SK | 792 | 7.49 | 7.43 | 7.47 | 7.43 | 8.32 | 8.36 | 8.35 | 8.10 |
| cs_cz | cs-CZ | 723 | 7.58 | 7.63 | 7.57 | 7.62 | 8.24 | 8.19 | 8.25 | 8.06 |
| sv_se | sv-SE | 759 | 7.99 | 7.99 | 7.95 | 8.03 | 7.98 | 7.87 | 7.93 | 8.02 |
| hr_hr | hr-HR | 914 | 8.22 | 8.26 | 8.21 | 8.28 | 9.31 | 9.26 | 9.26 | 9.43 |
| vi_vn | vi-VN | 857 | 8.84 | 8.81 | 8.83 | 8.88 | 9.03 | 8.98 | 9.07 | 9.06 |
| ko_kr | ko-KR | 382 | 9.18 | 9.20 | 9.23 | 9.25 | 9.28 | 9.39 | 9.28 | 9.30 |
| hu_hu | hu-HU | 905 | 9.98 | 9.93 | 9.91 | 10.06 | 10.59 | 10.42 | 10.40 | 10.69 |
| ro_ro | ro-RO | 883 | 10.02 | 10.02 | 9.95 | 9.78 | 12.09 | 12.08 | 12.06 | 11.21 |
| da_dk | da-DK | 930 | 10.88 | 10.82 | 10.81 | 10.88 | 10.96 | 10.86 | 10.87 | 10.83 |
| ja_jp | ja-JP | 650 | 13.89 | 13.85 | 13.82 | 13.63 | 13.91 | 13.88 | 13.95 | 13.77 |
| lt_lt | lt-LT | 986 | 14.61 | 14.59 | 14.59 | 14.74 | 16.36 | 16.39 | 16.53 | 17.15 |
| lv_lv | lv-LV | 851 | 14.97 | 14.93 | 14.95 | 15.03 | 15.42 | 15.35 | 15.40 | 15.51 |
| sl_si | sl-SI | 834 | 15.04 | 14.93 | 15.10 | 15.81 | 19.75 | 19.53 | 19.98 | 21.48 |
| el_gr | el-GR | 650 | 15.53 | 15.56 | 15.54 | 15.70 | 16.70 | 16.66 | 16.84 | 16.90 |
| mt_mt | mt-MT | 926 | 17.70 | 17.25 | 17.72 | 17.64 | 18.19 | 17.80 | 18.21 | 17.91 |
| cmn_hans_cn | zh-CN | 945 | 24.58 | 25.91 | 24.68 | 24.75 | 25.18 | 26.89 | 25.27 | 25.53 |
| he_il | he-IL | 792 | 28.60 | 28.72 | 28.54 | 27.54 | 33.37 | 33.77 | 33.54 | 33.08 |
| **MEAN (39 rows)** | | | **8.83** | **8.84** | **8.81** | **8.83** | **9.48** | **9.51** | **9.48** | **9.52** |

### Reading the table

**Parity.** All four explicit configurations land in 8.81–8.84, and all four auto configurations in
9.48–9.52. Against `ref-pd` — the matched-policy column — no locale degrades by more than 0.15, and
ours is ahead on 13 of the 34 independent locales.

**Explicit language beats `auto` by ~0.76 CER on average**, concentrated in lower-resource locales
(sl-SI +4.6, he-IL +5.1, hr-HR +1.0) and negligible for high-resource ones.

**bfloat16 costs 1.33 CER on Mandarin** (zh-CN 25.91 → 24.58 at fp32). No other locale moves by more
than ~0.2. This is consistent with more than half of the 13,087-token vocabulary being CJK, where
bf16's 8-bit mantissa blurs near-tied logits. Since the pipeline defaults to bf16, it is worth knowing
for zh-CN workloads.

**Verification.** 0 language tags and 0 stray angle brackets across 114,000 utterances — auto mode
previously leaked 6. The incomplete-tag fix changed exactly 6 utterances (he_il 5, sl_si 1), leaving
the other 26,733 byte-identical. Both implementations are fully deterministic across independent runs.
148 unit tests pass; isort and black clean.

### Caveats

- **Thai (th-TH) is excluded from the table.** It produces 82–93% empty output in *both*
  implementations (CER 100–134) — a model limitation consistent with its Adaptation-Ready tier.
  Including it would add roughly 2.6 to every mean and shift the comparison ~0.3 based purely on how
  empty-versus-garbage output scores.
- **Absolute CER is pessimistic by roughly 2.4 points.** ITN is off, so the model verbalizes numbers
  ("one fifty") while FLEURS references use digits ("150"). That costs 12–14 CER on the ~19% of
  utterances containing numbers. It affects both implementations identically, so the comparison is
  unaffected — but for example German's true non-numeric CER is 1.91, not 4.81.
  

## Language-mismatch behaviour

What happens when the audio is in language X but the prompt is set to a different locale Y — the
practical risk when manifest language metadata is stale or wrong. Same settings as above
(bf16, `att_context_size=[56,13]`, `batch_size=128`); `matched` and `auto` are the baselines from
the main table.

| # | audio X | prompt Y | relationship | utts | matched | auto | mismatched | Δ vs matched | Δ vs auto | empty % |
|---|---|---|---|---|---|---|---|---|---|---|
| 1 | Russian `ru_ru` | `bg-BG` | sibling Cyrillic | 775 | 5.26 | 5.49 | **5.44** | **+0.17** | −0.05 | 0.0 |
| 2 | Croatian `hr_hr` | `sl-SI` | sibling Slavic | 914 | 8.26 | 9.26 | **9.07** | **+0.80** | −0.19 | 0.0 |
| 3 | Czech `cs_cz` | `sk-SK` | sibling Slavic | 723 | 7.63 | 8.19 | **8.44** | **+0.81** | +0.25 | 0.0 |
| 4 | Danish `da_dk` | `sv-SE` | sibling Nordic | 930 | 10.82 | 10.86 | **13.99** | **+3.17** | +3.13 | 0.1 |
| 5 | German `de_de` | `nl-NL` | sibling Germanic | 862 | 4.78 | 4.98 | **17.13** | **+12.36** | +12.15 | 8.7 |
| 6 | Spanish `es_419` | `pt-BR` | sibling Romance | 908 | 3.57 | 3.58 | **24.76** | **+21.19** | +21.19 | 9.0 |
| 7 | French `fr_fr` | `tr-TR` | unrelated, Latin script | 676 | 5.27 | 5.28 | **60.16** | **+54.89** | +54.88 | 43.6 |
| 8 | Japanese `ja_jp` | `de-DE` | unrelated, other script | 650 | 13.85 | 13.88 | **99.79** | **+85.94** | +85.91 | 58.8 |
| 9 | English `en_us` | `zh-CN` | unrelated, other script | 647 | 5.89 | 6.54 | **99.78** | **+93.90** | +93.24 | 90.9 |
| 10 | German `de_de` | `ja-JP` | unrelated, other script | 862 | 4.78 | 4.98 | **99.98** | **+95.20** | +95.00 | 92.6 |

**The prompt is a hard constraint, not a soft hint.** Degradation scales with linguistic distance
across four orders of magnitude, from +0.17 CER to +95.20.
**The failure mode is silence, not mistranscription.**
**`auto` is the right default when language metadata is uncertain.**

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/inference/streaming/text/text_processing.py [hunk 1]
```diff
@@ -30,6 +30,12 @@
 if TYPE_CHECKING:
     from nemo.collections.asr.inference.itn.inverse_normalizer import AlignmentPreservingInverseNormalizer
 
+# An unterminated language tag, e.g. "<mt-MT" or "<sl-", left when the stream ends before the model
```
Comment (andrusenkoau) [inline nemo/collections/asr/inference/streaming/text/text_processing.py]: Could you clarify the problem with incomplete language tags? Why can the model predict it?

**Review by andrusenkoau — COMMENTED** [review]
The PR looks good. I left one question regarding to incomplete language tags. 

It is also interesting how the model behaves in the case of an incorrect language tag during inference.

File: nemo/collections/asr/inference/streaming/text/text_processing.py [hunk 2]
```diff
@@ -285,13 +295,17 @@ def generate_final_transcript(
...
                 state.final_segments.append(word.copy())
                 state.final_transcript += word.text + self.sep
             state.final_transcript = state.final_transcript.rstrip(self.sep)
 
         # Generate final transcript for segment boundary states
         for state in segment_boundary_states:
             for segment in state.segments:
-                state.final_segments.append(segment.copy())
+                segment = segment.copy()
+                segment.text = INCOMPLETE_LANG_TAG.sub("", segment.text)
```
Comment (andrusenkoau) [inline nemo/collections/asr/inference/streaming/text/text_processing.py]: Apply only to models with promt id support

**Review by andrusenkoau — APPROVED** [review]
Thanks for the update, LGTM.
